### ISSUE TITLE
Pushing ImGuiCol_WindowBg before Begin() is ignored if window is docked

### ISSUE THREAD

**Issue by tksuoran** [ocornut/imgui#5634]
**Version/Branch of Dear ImGui:**

Version: 1.89 WIP
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: custom/erhe
Operating System: Windows 10

**My Issue/Question:**

It appears that `ImGui::PushStyleColor(ImGuiCol_WindowBg)` is ignored if window is docker.

Background:
- I have set my desktop wallpaper to checkerboard
- I have enabled window transparency (`glfwWindowHint(GLFW_TRANSPARENT_FRAMEBUFFER, GLFW_TRUE);`)
- My goal is to have transparency in _one ImGui window_, specifically, a viewport window that shows 3D scene. Everything else should be opaque. To achieve this, I have opaque default ImGuiCol_WindowBg color for most windows, and I push transparent black (0, 0, 0, 0) for the viewport window
- 🟢 The effect works correctly when scene viewport window is not docked
- 🔴 The effect does not work when the scene viewport window is docked

**Screenshots**

🟢 Not docked, rendering as expected:
![not_docked](https://user-images.githubusercontent.com/994606/188152263-dd153a87-6efb-4cd2-9007-8a87a9c50775.png)

🔴 Docked, pushed ImGuiCol_WindowBg appears to be ignored
![docked](https://user-images.githubusercontent.com/994606/188153593-853d1ce5-1f58-4932-a797-63326a9bcea3.png)

```

**Comment by ocornut** [comment]
Also see #2700 and #2766
Can you try using (255,255,255,0) and confirm if it is that Alpha is ignored in that situation?

It's confusing what part of your issue are related to Docking and which part are related to Viewports. Mentioning `glfwWindowHint(GLFW_TRANSPARENT_FRAMEBUFFER, GLFW_TRUE);` suggest the later. 
Do you have the same issue if you disable multi-viewports?

**Comment by tksuoran** [comment]
(255,255,255,0) - or should it be (1,1,1,0) - I tried both - all result exactly the same.
How should I disable multi-viewports? I am not using multi-viewports (as far as I know).

**Comment by ocornut** [comment]
By not enabling `ImGuiConfigFlags_ViewportsEnable`. You can clear it in `Demo->Configuration`.

**Comment by tksuoran** [comment]
I don't enable `ImGuiConfigFlags_ViewportsEnable`. The only bit I set in `io.ConfigFlags` is `ImGuiConfigFlags_NavNoCaptureKeyboard`.

This may already be covered by the issues you pointed to - thanks, I will check them in more detail. Meanwhile, I did make the following observation, comparing ImGui draw commands with RenderDoc:

1. When all windows are docked, the whole ImGui viewport is covered with a rectangle, filling it with WindowBg color
2. When some window is docked, this covering of whole ImGui viewport does not happen, and individual windows are rendered directly to the framebuffer

In the case 1) this covering of the whole viewport appears to be the root cause why I cannot achieve the effect of having one window transparent if all windows are docked.

In other words: The operation of pushing transparent window background color to transparent black works fine. The issue is that by the time the window is rendered with opaque background color, ImGui viewport is already filled with opaque color.

**Comment by tksuoran** [comment]
I made a several edits to my previous comment, but I let it be now, promise.

One further info: I am using `ImGui::DockSpaceOverViewport(nullptr, ImGuiDockNodeFlags_PassthruCentralNode);`

Related: #3924

**Comment by tksuoran** [comment]
In `ImGui::DockNodeUpdate()` I commented out the code in section `// Draw whole dockspace background if ImGuiDockNodeFlags_PassthruCentralNode if set.` (typo there? 'is set' perhaps) and with that change I can achieve the effect I was looking for.

Why why would we need to draw the background there? That feels backwards how I understand 'passthru'.

**Comment by ocornut** [comment]
> Why why would we need to draw the background there? That feels backwards how I understand 'passthru'.

Notice it is calling `RenderRectFilledWithHole()`. Central node is passthrough, other node still have their background.

**Comment by tksuoran** [comment]
How is the central node related to the ImGui windows I have docked?

I have one or more of these scene viewport windows, and all of them should be rendered without any extra background. User may dock and undock them anywhere.

**Comment by ocornut** [comment]
(Sorry no time to look at this carefully now but will eventually!)

**Comment by ravikul505** [comment]
hello sir , i want to work on this

**Comment by ravikul505** [comment]
> (Sorry no time to look at this carefully now but will eventually!)

Hello @ocornut  , i want to work on this

**Comment by tvoeroes** [comment]
FYI I'm using this simple change to achieve transparent background: [transparency.patch](https://github.com/ocornut/imgui/files/13754612/transparency.patch)

**Comment by Ishisoft** [comment]
Inspired by the patch by @tvoeroes above, I'm using this similar change to allow just one of my windows to remain transparent while docked. If the window has the NoBackground flag it'll remain transparent while docked, while other windows keep their background. I'm not all that familiar with imgui code and the intricacies of rendering docked windows but it works for my purposes at least, and may be useful for someone. "&& (host_window->Flags & ImGuiWindowFlags_NoBackground) == 0" is the new bit.

    // Draw whole dockspace background if ImGuiDockNodeFlags_PassthruCentralNode if set.
    // We need to draw a background at the root level if requested by ImGuiDockNodeFlags_PassthruCentralNode, but we will only know the correct pos/size
    // _after_ processing the resizing splitters. So we are using the DrawList channel splitting facility to submit drawing primitives out of order!
    const bool render_dockspace_bg = node->IsRootNode() && host_window && (node_flags & ImGuiDockNodeFlags_PassthruCentralNode) != 0 && (host_window->Flags & ImGuiWindowFlags_NoBackground) == 0;
    if (render_dockspace_bg && node->IsVisible)
    {

**Comment by tracefree** [comment]
@Ishisoft Hey, I use your change in a fork and it works really well, thank you! Would you like to open a PR for it?

**Comment by YanisBDF** [comment]
Workaround that works for me :
Instead of setting ImGuiWindowFlags_NoBackground on the host dockspace window, I set its ImGuiCol_WindowBg to fully transparent and only use NoBackground on my central/content window.
Important: don’t set ImGuiWindowFlags_NoBackground on the host dockspace window—if you do, the host’s WindowBg color/alpha is ignored by design. Using PassthruCentralNode + transparent host bg lets the central area show through as expected.
